### ISSUE TITLE
Alpha Transparency Artifacts On TBitmap Assign

### ISSUE THREAD

**Issue by FMXExpress** [skia4delphi/skia4delphi#89]
I have a drag and drop system with a TPaintBox. The images I am dragging have transparency (probably PNG). With just GlobalUseMetal it doesn't have artifacts. With GlobalUseSkia it has artifacts.

DragImage is a TImage and floats over my form. I assign the image to my floating image from a TListBoxItem(Sender).ItemData.Bitmap.

 ```delphi
  DragImage.Bitmap.Clear(0); // required by Skia otherwise the image from previous use is under the new Assign()
  DragImage.Bitmap.Assign(TListBoxItem(Sender).ItemData.Bitmap); // I think the image corruption happens here with the transparency
  DragImage.Bitmap.Resize(50,50);
  DragImage.Position.X := X+IconsListBox.Position.X-Trunc(DragImage.Bitmap.Width/2);
  DragImage.Position.Y := TListBoxItem(Sender).Position.Y+Y+IconsListBox.Position.Y-Trunc(DragImage.Bitmap.Height/2);
  DragImage.Tag := TListBoxItem(Sender).Tag;
  DragImage.Visible := True;
  DragImage.BringToFront;
```

And this is what my stamp down onto the TPaintBox looks like.

```delphi
procedure TMyPaintBox.StampBitmap(X,Y: Single; B: TBitmap);
var
r,rd:TRectF;
begin
  with fdrawbmp.Canvas do
  begin
  BeginScene();
    if (assigned(B)) then
     begin
      r:=TRectF.Create(PointF(0,0),B.Width,B.Height);
      rd:=TRectF.Create(PointF(X-(B.Width/2),Y-(B.Height/2)),B.Width,B.Height);
      DrawBitmap(B,r,rd,1);
     end;
  EndScene;
  end;
  forigbmp.Assign(fdrawbmp);
  {$IFDEF POSIX}
  InvalidateRect(fdrawrect);
  {$ENDIF}
end;
```

![20220326_065845](https://user-images.githubusercontent.com/7482023/160243839-97589d4e-bb0f-43b5-b7e7-b184ed7d539a.jpg)

**Comment by viniciusfbb** [comment]
Hi @FMXExpress, thank you very much for reporting the issue!

> DragImage.Bitmap.Clear(0); // required by Skia otherwise the image from previous use is under the new Assign()

Before a Bitmap.Assign, no clear is required. If in any case this is really necessary, it is an issue and should be reported.

About your drawing code, there is a problem. TBitmap almost always starts clean with FMX's default Canvas, but it's not guaranteed, so you need a Bitmap.Canvas.Clear(TAlphaColors.Null) after BeginScene if you don't make an opaque drawing, which is the case. In Skia's Canvas this problem is clearer because the bitmap will always start with garbage.

So in either case, but especially when using skia, use Bitmap.Canvas.Clear(0) after BeginScene.

**Comment by viniciusfbb** [comment]
@FMXExpress Did you solve it by adding the clear? Can I close the issue?

**Comment by FMXExpress** [comment]
I haven't tried to solve it yet based on your feedback. Also not clear what your code suggestion is. Goal was to get more screenshots before responding.

My theory is that the DragImage (a TImage) is where the corruption happens.

Without the following line the DragImage gets each dragged icon stacked up inside it (because they are transparent you can see the old one). And that doesn't happen with FMX or Metal canvases.
DragImage.Bitmap.Clear(0); // required by Skia otherwise the image from previous use is under the new Assign()

It is also possible that the corruption is from TListBoxItem(Sender).ItemData.Bitmap. I haven't looked at it's loading code. I would assume they load a PNG though.

**Comment by viniciusfbb** [comment]
> Also not clear what your code suggestion is.

My suggestion was to just add a clear to the bitmap before starting to draw on it.

```delphi
procedure TMyPaintBox.StampBitmap(X,Y: Single; B: TBitmap);
var
r,rd:TRectF;
begin
  with fdrawbmp.Canvas do
  begin
  BeginScene();
  fdrawbmp.Canvas.Clear(TAlphaColors.Null); // <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
    if (assigned(B)) then
     begin
      r:=TRectF.Create(PointF(0,0),B.Width,B.Height);
      rd:=TRectF.Create(PointF(X-(B.Width/2),Y-(B.Height/2)),B.Width,B.Height);
      DrawBitmap(B,r,rd,1);
     end;
  EndScene;
  end;
  forigbmp.Assign(fdrawbmp);
  {$IFDEF POSIX}
  InvalidateRect(fdrawrect);
  {$ENDIF}
end;
```

**Comment by viniciusfbb** [comment]
@FMXExpress We decided to start the bitmap already clean to avoid this kind of problem.

Fixed in new release [v3.3.1](https://github.com/skia4delphi/skia4delphi/releases)

**Comment by FMXExpress** [comment]
Confirming this fixed the issue.
